File: binarysearch.io/November 2020/211_level_order_alternating.py
# class Tree:
#     def __init__(self, val, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque 
class Solution:
    def solve(self, root):
        s = 0
        q = []
        q.append(root)
        ans = [root.val,]
        while len(q):
            new = []
            for node in q:
                if node.left:
                    new.append(node.left)
                if node.right:
                    new.append(node.right)
            if s%2:
                ans.extend([node.val for node in new])
            else:
                ans.extend([node.val for node in reversed(new)])
            q = new
            s += 1
        return ans

File: binarysearch.io/211_level_order_alternating.cpp
/**
 * class Tree {
 *     public:
 *         int val;
 *         Tree *left;
 *         Tree *right;
 * };
 */
void add_to_list(int s,vector<int>  & order, vector<Tree *>  newl){
    if(s){
        for(int i = newl.size()-1; i >=0; i--){
            order.push_back(newl[i]->val);
        }
    }else{
        for(int i = 0; i < newl.size(); i++){
            order.push_back(newl[i]->val);
        }
    }
}
vector<int> solve(Tree* root) {
    vector<Tree *> stack;
    stack.push_back(root);
    int s = 1;
    vector<int> order;
    order.push_back(root->val);
    while(stack.size()){
        vector<Tree *> newl;
        for(auto &node : stack){
            if(node->left) newl.push_back(node->left);
            if(node->right) newl.push_back(node->right);
        }
        add_to_list(s,order,newl);
        stack = newl;
        s ^= 1;
    }
    return order;
}